### PR TITLE
Added an optional config in the CR to specify Flink application service account

### PR DESCRIPTION
**Pull request by yuchaoran2011**
#76

### CODE REVIEW

**Comment by yuchaoran2011** [comment]
Looks like the unit test failed because the hash changed. But adding an optional field in the CR with a default value isn't an API-breaking change, or is it?

**Comment by anandswaminathan** [comment]
Currently it is. To make it non breaking, you will have to set serviceAccount only if the serviceAccount is set in the flink application.

**Comment by yuchaoran2011** [comment]
Got it. Now the tests are fixed as suggested.

**Comment by mwylde** [comment]
Hey @yuchaoran2011 — is this something you could set using a mutating admission controller webhook? I'm a bit leery of going down the road of making every deployment / pod / service option configurable via the FlinkApplication resource.

**Comment by yuchaoran2011** [comment]
I haven't tried but that sounds like much extra work to satisfy a use case that is as common as setting the service account. And I'm not "making every deployment/pod/service" configurable, just the **service accounts of the application JM and TM deployments**, not the operator deployment and not any other resource kind.

**Comment by anandswaminathan** [comment]
Agree with @yuchaoran2011. ‘ServiceAccount’ is similar to other fields we already have in the custom resource like volume, env, ImagePullSecrets, etc. It fits in well with general kubernetes pattern. Every resource has a serviceAccount. Flink application is just another resource like a deployment or cron.

File: pkg/apis/app/v1beta1/types.go [hunk 1]
```diff
@@ -33,6 +33,7 @@ type FlinkApplicationSpec struct {
 	Image                 string                       `json:"image,omitempty" protobuf:"bytes,2,opt,name=image"`
 	ImagePullPolicy       apiv1.PullPolicy             `json:"imagePullPolicy,omitempty" protobuf:"bytes,14,opt,name=imagePullPolicy,casttype=PullPolicy"`
 	ImagePullSecrets      []apiv1.LocalObjectReference `json:"imagePullSecrets,omitempty" patchStrategy:"merge" patchMergeKey:"name" protobuf:"bytes,15,rep,name=imagePullSecrets"`
+	ServiceAccount        string                       `json:"serviceAccount,omitempty"`
```
Comment (anandswaminathan) [inline pkg/apis/app/v1beta1/types.go]: Can we call it serviceAccountName - consistent with other kubernetes resources?

File: pkg/controller/flink/config.go [hunk 2]
```diff
@@ -42,6 +43,10 @@ func getJobmanagerReplicas(app *v1beta1.FlinkApplication) int32 {
 	return firstNonNil(app.Spec.JobManagerConfig.Replicas, JobManagerDefaultReplicaCount)
 }
 
+func getFlinkSa(app *v1beta1.FlinkApplication) string {
+	return app.Spec.ServiceAccount
```
Comment (anandswaminathan) [inline pkg/controller/flink/config.go]: Rename to getServiceAccountName ?

File: pkg/apis/app/v1beta1/types.go (hunk 1 above)
Comment (yuchaoran2011) [reply pkg/apis/app/v1beta1/types.go]: Done.

**Comment by yuchaoran2011** [comment]
The integration test failed with error "the job exceeded maximum log depth". Not sure what happened. If one of the admins can restart the integration test, maybe the error will be gone.

**Comment by mwylde** [comment]
Thanks @anandswaminathan for the explanation. That sounds reasonable.

@yuchaoran2011 unfortunately, that's just flakiness in the integration tests. It should pass if you re-run it (and we need to spend some time getting the tests reliable again).

**Comment by yuchaoran2011** [comment]
Yeah I guessed so. But I don't think I have the permission to re-run the tests. If any one of you could do it, that would be great.

File: pkg/controller/flink/job_manager_controller.go [hunk 3]
```diff
@@ -335,6 +335,12 @@ func jobmanagerTemplate(app *v1beta1.FlinkApplication) *v1.Deployment {
 			},
 		},
 	}
+
+	flinkSa := getServiceAccountName(app)
```
Comment (anandswaminathan) [inline pkg/controller/flink/job_manager_controller.go]: Nit: rename flinkSa --> serviceAccountName

**Comment by mwylde** [comment]
I've restarted the job, but it's unfortunate that you weren't able to yourself. I'll look into getting that fixed.

**Comment by anandswaminathan** [comment]
@yuchaoran2011 Merging this.

**Pull request merged**